lib/msal-core/src/utils/WindowUtils.ts

```diff
@@ -0,0 +1,202 @@
+        return new Promise((resolve, reject) => {
+            const maxTicks = timeout / WindowUtils.POLLING_INTERVAL_MS;
+            let ticks = 0;
+
+            const intervalId = setInterval(() => {
+                let href;
+                try {
+                    /*
+                     * Will throw if cross origin,
+                     * which should be caught and ignored
+                     * since we need the interval to keep running while on STS UI.
+                     */
+                    href = contentWindow.location.href;
+                } catch (e) {}
+
+                // Don't process blank pages or cross domain
+                if (!href || href === "about:blank") {
+                    return;
+                }
+
+                // Only run clock when we are on same domain
+                ticks++;
+
+                if (UrlUtils.urlContainsHash(href)) {
+                    clearInterval(intervalId);
+                    resolve(contentWindow.location.hash);
+                } else if (contentWindow.closed) {
+                    clearInterval(intervalId);
+                    resolve();
+                } else if (ticks > maxTicks) {
+                    clearInterval(intervalId);
+                    reject(ClientAuthError.createTokenRenewalTimeoutError()); // better error?
+                }
+            }, WindowUtils.POLLING_INTERVAL_MS);
+        });
+    }
+
+    /**
+     * @hidden
+     * Loads iframe with authorization endpoint URL
+     * @ignore
+     */
+    static loadFrame(urlNavigate: string, frameName: string, timeoutMs: number, logger: Logger): Promise<HTMLIFrameElement> {
+        /*
+         * This trick overcomes iframe navigation in IE
+         * IE does not load the page consistently in iframe
+         */
+        logger.info("LoadFrame: " + frameName);
+
+        return new Promise((resolve, reject) => {
+            setTimeout(() => {
+                const frameHandle = WindowUtils.addHiddenIFrame(frameName, logger);
+
+                if (!frameHandle) {
+                    reject(`Unable to load iframe with name: ${frameName}`);
+                    return;
+                }
+
+                if (frameHandle.src === "" || frameHandle.src === "about:blank") {
+                    frameHandle.src = urlNavigate;
+                    logger.infoPii("Frame Name : " + frameName + " Navigated to: " + urlNavigate);
+                }
+
+                resolve(frameHandle);
+            }, timeoutMs);
+        });
+    }
+
+    /**
+     * @hidden
+     * Adds the hidden iframe for silent token renewal.
+     * @ignore
+     */
+    static addHiddenIFrame(iframeId: string, logger: Logger): HTMLIFrameElement {
+        if (typeof iframeId === "undefined") {
+            return null;
+        }
+
+        logger.info("Add msal frame to document:" + iframeId);
+        let adalFrame = document.getElementById(iframeId) as HTMLIFrameElement;
```

Comment:
i know you didnt write this, but can we rename the `adalFrame` it has nothing to do with adal and adds a lot of confusion

Reply:
I'm going to come back later and clean up this function, for now going to keep it as-is since its copy/pasta.